{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "229eb4ac-73c1-41a3-b72e-0fcb5dacbeb7",
   "metadata": {},
   "source": [
    "# data loading"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "6ee427c8-c8a5-4983-965f-3777c64f904e",
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "20782bb2-d79a-48a1-bd21-e068dacaf721",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/christoffer/miniconda3/envs/sc/lib/python3.8/site-packages/anndata/_core/anndata.py:1838: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"obs\")\n"
     ]
    }
   ],
   "source": [
    "adata = sc.read_h5ad('../../data/RREAE_5k_raw_integration_processed.h5ad')\n",
    "adata.obs.index = adata.obs.index.astype(str)\n",
    "adata.obs_names_make_unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "59d2d863-76db-4d6f-84b7-27da7ff0cbc0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_id</th>\n",
       "      <th>sample_name</th>\n",
       "      <th>region</th>\n",
       "      <th>course</th>\n",
       "      <th>condition</th>\n",
       "      <th>model</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>S1-B1_1</td>\n",
       "      <td>R1-1_T</td>\n",
       "      <td>T</td>\n",
       "      <td>remitt I</td>\n",
       "      <td>EAE</td>\n",
       "      <td>RR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>S1-B1_0</td>\n",
       "      <td>OS1-1_L</td>\n",
       "      <td>L</td>\n",
       "      <td>onset I</td>\n",
       "      <td>EAE</td>\n",
       "      <td>RR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>S1-B1_2</td>\n",
       "      <td>P2-1_C</td>\n",
       "      <td>C</td>\n",
       "      <td>peak I</td>\n",
       "      <td>EAE</td>\n",
       "      <td>RR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>S1-B2_2</td>\n",
       "      <td>P2-1_L</td>\n",
       "      <td>L</td>\n",
       "      <td>peak II</td>\n",
       "      <td>EAE</td>\n",
       "      <td>RR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>S1-B2_0</td>\n",
       "      <td>OS1_1_T</td>\n",
       "      <td>T</td>\n",
       "      <td>onset I</td>\n",
       "      <td>EAE</td>\n",
       "      <td>RR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>102</th>\n",
       "      <td>G6_L3_0</td>\n",
       "      <td>872025 TSC</td>\n",
       "      <td>C</td>\n",
       "      <td>chronic long</td>\n",
       "      <td>EAE</td>\n",
       "      <td>Chronic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>103</th>\n",
       "      <td>G6_L3_2</td>\n",
       "      <td>872021 CTSC-C</td>\n",
       "      <td>C</td>\n",
       "      <td>MOG CFA</td>\n",
       "      <td>CONTROL</td>\n",
       "      <td>Chronic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>104</th>\n",
       "      <td>G4_L1_1</td>\n",
       "      <td>872025 LSC</td>\n",
       "      <td>L</td>\n",
       "      <td>chronic long</td>\n",
       "      <td>EAE</td>\n",
       "      <td>Chronic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>105</th>\n",
       "      <td>G4_L1_0</td>\n",
       "      <td>872022 LSC</td>\n",
       "      <td>L</td>\n",
       "      <td>chronic long</td>\n",
       "      <td>EAE</td>\n",
       "      <td>Chronic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>106</th>\n",
       "      <td>G4_L1_2</td>\n",
       "      <td>872021 LSC</td>\n",
       "      <td>L</td>\n",
       "      <td>MOG CFA</td>\n",
       "      <td>CONTROL</td>\n",
       "      <td>Chronic</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>107 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    sample_id    sample_name region        course condition    model\n",
       "0     S1-B1_1         R1-1_T      T      remitt I       EAE       RR\n",
       "1     S1-B1_0        OS1-1_L      L       onset I       EAE       RR\n",
       "2     S1-B1_2         P2-1_C      C        peak I       EAE       RR\n",
       "3     S1-B2_2         P2-1_L      L       peak II       EAE       RR\n",
       "4     S1-B2_0        OS1_1_T      T       onset I       EAE       RR\n",
       "..        ...            ...    ...           ...       ...      ...\n",
       "102   G6_L3_0     872025 TSC      C  chronic long       EAE  Chronic\n",
       "103   G6_L3_2  872021 CTSC-C      C       MOG CFA   CONTROL  Chronic\n",
       "104   G4_L1_1     872025 LSC      L  chronic long       EAE  Chronic\n",
       "105   G4_L1_0     872022 LSC      L  chronic long       EAE  Chronic\n",
       "106   G4_L1_2     872021 LSC      L       MOG CFA   CONTROL  Chronic\n",
       "\n",
       "[107 rows x 6 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "anno = pd.read_csv('/Users/christoffer/Downloads/2025_Xenium5k_sample_annotation_integration_TS.csv', sep = ';', )\n",
    "anno"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "1d256c8d-16b9-49cb-a787-a36e5ce4408d",
   "metadata": {},
   "outputs": [],
   "source": [
    "for meta in ['sample_name', 'region', 'course','condition','model']:\n",
    "    mapping_dict = dict(zip(anno['sample_id'], anno[meta]))\n",
    "    adata.obs[meta] = adata.obs['sample_id'].map(mapping_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "d1900bee-fb40-4ebc-8757-b6448538c9ff",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.layers['raw'] = adata.layers['raw'].astype(int)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "070da72e-4491-40e2-ad2a-398c6c60ca54",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.X = adata.layers['raw']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "a70c1c63-b940-4027-af34-9f496d1f6222",
   "metadata": {},
   "outputs": [],
   "source": [
    "del adata.layers['raw']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "70e6487e-76a6-457c-afcb-fb6815b46484",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 891821 × 5101\n",
       "    obs: 'x_centroid', 'y_centroid', 'transcript_counts', 'control_probe_counts', 'genomic_control_counts', 'control_codeword_counts', 'unassigned_codeword_counts', 'deprecated_codeword_counts', 'total_counts', 'cell_area', 'nucleus_area', 'nucleus_count', 'segmentation_method', 'run', 'grid_label', 'project', 'strain', 'n_genes_by_counts', 'n_counts', 'n_genes', 'leiden_0.5', 'leiden_1', 'leiden_1.5', 'leiden_2', 'leiden_2.5', 'sample_id', 'sub_type', 'cell_type', 'cell_class', 'area', 'sample_name', 'region', 'course', 'condition', 'model', 'cell_id', 'rbd_domain', 'sub_type_I', 'sub_type_II', 'sub_type_III', 'cell_id_2', 'match_key', 'lesion_density_call', 'lesion_distance_um', 'lesion_distance_bin', 'celltype_merged', 'dist_bin'\n",
       "    var: 'n_cells_by_counts', 'mean_counts', 'pct_dropout_by_counts', 'total_counts', 'mean', 'std'\n",
       "    uns: 'cell_class_colors', 'cell_polygons', 'cell_type_colors', 'condition_colors', 'course_colors', 'dendrogram_sub_type', 'dendrogram_sub_type_II', 'leiden', 'leiden_0.5_colors', 'leiden_1.5_colors', 'leiden_1_colors', 'leiden_2.5_colors', 'leiden_2_colors', 'log1p', 'model_colors', 'neighbors', 'pca', 'rank_genes_groups', 'rbd_domain_colors', 'region_colors', 'sample_name_colors', 'sub_type_III_colors', 'sub_type_II_colors', 'sub_type_II_t_test', 'sub_type_I_colors', 'sub_type_colors', 'umap'\n",
       "    obsm: 'X_pca', 'X_umap', 'spatial'\n",
       "    varm: 'PCs'\n",
       "    obsp: 'connectivities', 'distances'"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "a798b367-c491-49dd-b9c4-791a6d9021e6",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.write('../../data/RREAE_5k_raw_only_integration_processed.h5ad')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "9f768e8e-aa6e-49d1-8d24-4fd81811427c",
   "metadata": {},
   "outputs": [],
   "source": [
    "MODELS = {'MOG': {'baseline': 'MOG CFA',\n",
    "  'courses': ['non symptomatic',\n",
    "   'early onset',\n",
    "   'chronic peak',\n",
    "   'chronic long']},\n",
    " 'PLP': {'baseline': 'PLP CFA',\n",
    "  'courses': ['onset I',\n",
    "   'onset II',\n",
    "   'peak I',\n",
    "   'monophasic',\n",
    "   'remitt I',\n",
    "   'peak II',\n",
    "   'remitt II',\n",
    "   'peak III']}}\n",
    "\n",
    "genes = ['Hif1a','Hk2','Pfkl','Pdk1','Pkm','Ldha','Ldhb','Slc16a1','Slc16a3','Serpina3n','Ppargc1a',\"Mfn1\",\"Mfn2\",\"Opa1\",'Sirt2']\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "220afde6-67e6-40b0-8dfd-204f39691a23",
   "metadata": {},
   "source": [
    "# utilities"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "f728a709-22d1-4062-836d-572d9e01e9dc",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scipy.sparse as sp\n",
    "\n",
    "def _to_dense(X):\n",
    "    return X.toarray() if sp.issparse(X) else np.asarray(X)\n",
    "\n",
    "def get_matrix(adata, layer=None):\n",
    "    if layer is None:\n",
    "        return adata.X\n",
    "    if layer in adata.layers:\n",
    "        return adata.layers[layer]\n",
    "    raise ValueError(f\"Layer '{layer}' not found.\")\n",
    "\n",
    "def compute_logcpm(adata, layer=None):\n",
    "    X = _to_dense(get_matrix(adata, layer))\n",
    "    lib = X.sum(axis=1, keepdims=True)\n",
    "    lib[lib==0] = 1\n",
    "    return np.log1p((X / lib) * 1e6)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7e2f5285-5803-43bb-bed8-39274d378d78",
   "metadata": {},
   "source": [
    "# Function to compute variability"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "7399094f-69cb-40db-815f-b2c7a2fb5052",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scipy.sparse as sp\n",
    "\n",
    "def _to_dense(X):\n",
    "    return X.toarray() if sp.issparse(X) else np.asarray(X)\n",
    "\n",
    "def _get_matrix(adata, layer=None):\n",
    "    if layer is None:\n",
    "        return adata.X\n",
    "    if layer in adata.layers:\n",
    "        return adata.layers[layer]\n",
    "    raise ValueError(f\"Layer '{layer}' not found. Available: {list(adata.layers.keys())}\")\n",
    "\n",
    "def _compute_logcpm(adata, layer=None):\n",
    "    X = _to_dense(_get_matrix(adata, layer))\n",
    "    lib = X.sum(axis=1, keepdims=True); lib[lib == 0] = 1\n",
    "    return np.log1p((X / lib) * 1e6)\n",
    "\n",
    "def compute_dotplot_stats_by_proximity(\n",
    "    adata,\n",
    "    genes=None,                          # None = all genes\n",
    "    celltype_col=\"celltype_merged\",\n",
    "    proximity_col=\"lesion_proximity\",\n",
    "    counts_layer=\"raw\",\n",
    "    selected_celltypes=None,\n",
    "    selected_proximities=None,\n",
    "    min_cells_per_group=0\n",
    "):\n",
    "    \"\"\"\n",
    "    Returns tidy DataFrame with columns:\n",
    "      [celltype_merged, lesion_proximity, gene, mean_logcpm, pct_expressing]\n",
    "    \"\"\"\n",
    "\n",
    "    # ---- subset cells\n",
    "    mask = pd.Series(True, index=adata.obs_names)\n",
    "    if selected_celltypes is not None:\n",
    "        mask &= adata.obs[celltype_col].astype(str).isin(selected_celltypes)\n",
    "    if selected_proximities is not None:\n",
    "        mask &= adata.obs[proximity_col].astype(str).isin(selected_proximities)\n",
    "    if mask.sum() == 0:\n",
    "        raise ValueError(\"No cells left after filtering cell types / proximity.\")\n",
    "\n",
    "    ad = adata[mask].copy()\n",
    "    ad.obs_names_make_unique()\n",
    "    ad.obs[celltype_col]  = ad.obs[celltype_col].astype(str)\n",
    "    ad.obs[proximity_col] = ad.obs[proximity_col].astype(str)\n",
    "\n",
    "    # --- choose genes\n",
    "    if genes is None:\n",
    "        genes = ad.var_names.tolist()\n",
    "    else:\n",
    "        genes = [g for g in genes if g in ad.var_names]\n",
    "        if not genes:\n",
    "            raise ValueError(\"None of the requested genes are in adata.var_names.\")\n",
    "\n",
    "    # per-cell logCPM\n",
    "    logcpm = _compute_logcpm(ad, layer=counts_layer)\n",
    "    logcpm_df = pd.DataFrame(logcpm, index=ad.obs_names, columns=ad.var_names)[genes]\n",
    "\n",
    "    # long + metadata\n",
    "    long = (\n",
    "        logcpm_df.stack().rename(\"logcpm\").reset_index()\n",
    "        .rename(columns={\"level_0\":\"cell_id\",\"level_1\":\"gene\"})\n",
    "        .merge(\n",
    "            ad.obs[[celltype_col, proximity_col]].reset_index().rename(columns={\"index\":\"cell_id\"}),\n",
    "            on=\"cell_id\", how=\"left\"\n",
    "        )\n",
    "    )\n",
    "\n",
    "    # drop tiny groups\n",
    "    if min_cells_per_group and min_cells_per_group > 0:\n",
    "        sizes = long.groupby([celltype_col, proximity_col]).size()\n",
    "        keep = sizes[sizes >= min_cells_per_group].index\n",
    "        long = long.set_index([celltype_col, proximity_col]).loc[keep].reset_index()\n",
    "        if long.empty:\n",
    "            raise ValueError(\"All (celltype×proximity) groups dropped by min_cells_per_group.\")\n",
    "\n",
    "    # aggregate\n",
    "    stats = (\n",
    "        long.groupby([celltype_col, proximity_col, \"gene\"], observed=True)\n",
    "            .agg(mean_logcpm=(\"logcpm\",\"mean\"),\n",
    "                 pct_expressing=(\"logcpm\", lambda x: (x > 0).mean()*100.0))\n",
    "            .reset_index()\n",
    "    )\n",
    "\n",
    "    # lock ordering\n",
    "    stats[\"gene\"] = pd.Categorical(stats[\"gene\"], categories=genes, ordered=True)\n",
    "    if selected_proximities is not None:\n",
    "        stats[proximity_col] = pd.Categorical(\n",
    "            stats[proximity_col],\n",
    "            categories=selected_proximities,\n",
    "            ordered=True\n",
    "        )\n",
    "    if selected_celltypes is not None:\n",
    "        stats[celltype_col] = pd.Categorical(\n",
    "            stats[celltype_col],\n",
    "            categories=selected_celltypes,\n",
    "            ordered=True\n",
    "        )\n",
    "\n",
    "    return stats.sort_values([celltype_col, proximity_col, \"gene\"]).reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "e0d0e9ba-0c00-44ca-9edc-2fcc296b2f41",
   "metadata": {},
   "outputs": [],
   "source": [
    "def variable_genes_over_proximity(\n",
    "    stats,\n",
    "    celltype,\n",
    "    proximity_col=\"lesion_proximity\",\n",
    "    how=(\"logcpm\",\"pct\"),   # choose one or both; returns both by default\n",
    "):\n",
    "    \"\"\"\n",
    "    Rank genes by variability across proximity for a single cell type.\n",
    "    stats must have columns: [celltype_merged, proximity_col, gene, mean_logcpm, pct_expressing]\n",
    "    Returns a dict with 'logcpm' and/or 'pct' DataFrames sorted by variance.\n",
    "    \"\"\"\n",
    "    df = stats.loc[stats[\"celltype_merged\"].astype(str) == str(celltype)].copy()\n",
    "    if df.empty:\n",
    "        raise ValueError(f\"No rows for celltype='{celltype}' in stats.\")\n",
    "\n",
    "    # pivot to genes × proximity\n",
    "    mat_logcpm = df.pivot(index=\"gene\", columns=proximity_col, values=\"mean_logcpm\")\n",
    "    mat_pct    = df.pivot(index=\"gene\", columns=proximity_col, values=\"pct_expressing\")\n",
    "\n",
    "    out = {}\n",
    "    if \"logcpm\" in how:\n",
    "        var_log = mat_logcpm.var(axis=1, skipna=True)\n",
    "        rng_log = (mat_logcpm.max(axis=1) - mat_logcpm.min(axis=1))\n",
    "        out[\"logcpm\"] = (\n",
    "            pd.DataFrame({\"var\": var_log, \"range\": rng_log})\n",
    "              .sort_values(\"var\", ascending=False)\n",
    "        )\n",
    "    if \"pct\" in how:\n",
    "        var_pct = mat_pct.var(axis=1, skipna=True)\n",
    "        rng_pct = (mat_pct.max(axis=1) - mat_pct.min(axis=1))\n",
    "        out[\"pct\"] = (\n",
    "            pd.DataFrame({\"var\": var_pct, \"range\": rng_pct})\n",
    "              .sort_values(\"var\", ascending=False)\n",
    "        )\n",
    "    return {\"mat_logcpm\": mat_logcpm, \"mat_pct\": mat_pct, \"ranks\": out}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "a92ed691-30be-4446-9e3d-72cb576a1d8f",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "KeyboardInterrupt\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# define your proximity column + order\n",
    "proximity_col = \"lesion_distance_bin\"\n",
    "proximity_order = [\"0–10µm\",\"10–25µm\",\"25–50µm\",\"50–100µm\",\"100–200µm\",\"200–500µm\",\">500µm\"]  # change to your labels/order\n",
    "\n",
    "# compute stats for the cell types and genes you care about\n",
    "stats_prox = compute_dotplot_stats_by_proximity(\n",
    "    adata,\n",
    "    genes=None,\n",
    "    celltype_col=\"celltype_merged\",\n",
    "    proximity_col=proximity_col,\n",
    "    counts_layer=\"raw\",\n",
    "    selected_celltypes=[\"Astrocyte\", \"Oligodendrocyte\"],  # or any set\n",
    "    selected_proximities=proximity_order,\n",
    "    min_cells_per_group=20\n",
    ")\n",
    "\n",
    "# rank most variable genes over proximity for oligodendrocytes\n",
    "res = variable_genes_over_proximity(\n",
    "    stats_prox, celltype=\"Oligodendrocyte\", proximity_col=proximity_col, how=(\"logcpm\",\"pct\")\n",
    ")\n",
    "top_log = res[\"ranks\"][\"logcpm\"].head(15)\n",
    "top_pct = res[\"ranks\"][\"pct\"].head(15)\n",
    "print(\"Top by variance (mean logCPM):\\n\", top_log, \"\\n\")\n",
    "print(\"Top by variance (% expressing):\\n\", top_pct)\n",
    "\n",
    "# quick plot trajectories for the top 5 (mean logCPM)\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "\n",
    "top5 = top_log.head(10).index.tolist()\n",
    "palette = sns.color_palette(\"tab10\", n_colors=len(top5))\n",
    "\n",
    "mat = res[\"mat_logcpm\"].loc[top5, proximity_order]  # aligned to your order\n",
    "fig, ax = plt.subplots(figsize=(7,4))\n",
    "for g, col in zip(top5, palette):\n",
    "    ax.plot(proximity_order, mat.loc[g], marker=\"o\", label=fr\"$\\it{{{g}}}$\", color=col)\n",
    "ax.set_ylabel(\"mean logCPM\")\n",
    "ax.set_title(\"Oligodendrocyte — top variable genes across lesion proximity\")\n",
    "ax.legend(bbox_to_anchor=(1.02,1), loc=\"upper left\")\n",
    "plt.xticks(rotation=45)\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ac8460dc-b0cd-49ed-b2d2-26b53edeace7",
   "metadata": {},
   "source": [
    "# Plot top genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "76785782-c11e-4fb2-8c3e-386596447eae",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'variable_genes_per_model' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[18], line 5\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m \u001b[38;5;21;01mseaborn\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m \u001b[38;5;21;01msns\u001b[39;00m\n\u001b[1;32m      4\u001b[0m model \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mPLP\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[0;32m----> 5\u001b[0m cm, var \u001b[38;5;241m=\u001b[39m \u001b[43mvariable_genes_per_model\u001b[49m(\n\u001b[1;32m      6\u001b[0m     adata, model_name\u001b[38;5;241m=\u001b[39mmodel, MODELS\u001b[38;5;241m=\u001b[39mMODELS,\n\u001b[1;32m      7\u001b[0m     celltype\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mOligodendrocyte\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m      8\u001b[0m )\n\u001b[1;32m     10\u001b[0m top_genes \u001b[38;5;241m=\u001b[39m var\u001b[38;5;241m.\u001b[39mhead(\u001b[38;5;241m5\u001b[39m)\u001b[38;5;241m.\u001b[39mindex\n\u001b[1;32m     11\u001b[0m palette \u001b[38;5;241m=\u001b[39m sns\u001b[38;5;241m.\u001b[39mcolor_palette(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mtab10\u001b[39m\u001b[38;5;124m\"\u001b[39m, n_colors\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mlen\u001b[39m(top_genes))\n",
      "\u001b[0;31mNameError\u001b[0m: name 'variable_genes_per_model' is not defined"
     ]
    }
   ],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "\n",
    "model = \"PLP\"\n",
    "cm, var = variable_genes_per_model(\n",
    "    adata, model_name=model, MODELS=MODELS,\n",
    "    celltype=\"Oligodendrocyte\"\n",
    ")\n",
    "\n",
    "top_genes = var.head(5).index\n",
    "palette = sns.color_palette(\"tab10\", n_colors=len(top_genes))\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(8,5))\n",
    "for g, col in zip(top_genes, palette):\n",
    "    ax.plot(cm.columns, cm.loc[g], marker=\"o\", color=col, label=fr\"$\\it{{{g}}}$\")\n",
    "\n",
    "ax.set_ylabel(\"mean logCPM\")\n",
    "ax.set_title(f\"{model} — Astrocyte: Top variable genes\")\n",
    "ax.legend(bbox_to_anchor=(1.05,1), loc=\"upper left\")\n",
    "plt.xticks(rotation=45)\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "34c7656b-c49d-4e67-aa0b-5e30622907c0",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "\n",
    "model = \"PLP\"\n",
    "cm, var = variable_genes_per_model(\n",
    "    adata, model_name=model, MODELS=MODELS,\n",
    "    celltype=\"Astrocyte\"\n",
    ")\n",
    "\n",
    "top_genes = var.head(5).index\n",
    "palette = sns.color_palette(\"tab10\", n_colors=len(top_genes))\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(8,5))\n",
    "for g, col in zip(top_genes, palette):\n",
    "    ax.plot(cm.columns, cm.loc[g], marker=\"o\", color=col, label=fr\"$\\it{{{g}}}$\")\n",
    "\n",
    "ax.set_ylabel(\"mean logCPM\")\n",
    "ax.set_title(f\"{model} — Astrocyte: Top variable genes\")\n",
    "ax.legend(bbox_to_anchor=(1.05,1), loc=\"upper left\")\n",
    "plt.xticks(rotation=45)\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "98aa5d5e-130a-4976-a9a2-722a51b46a95",
   "metadata": {},
   "outputs": [],
   "source": [
    "var_sub = var[var.index.isin(genes)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f4ea191a-a101-44fc-b5e4-12edf5eb366e",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "\n",
    "model = \"PLP\"\n",
    "cm, var = variable_genes_per_model(\n",
    "    adata, model_name=model, MODELS=MODELS,\n",
    "    celltype=\"Astrocyte\"\n",
    ")\n",
    "\n",
    "top_genes = var_sub.head(5).index\n",
    "palette = sns.color_palette(\"tab10\", n_colors=len(top_genes))\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(8,5))\n",
    "for g, col in zip(top_genes, palette):\n",
    "    ax.plot(cm.columns, cm.loc[g], marker=\"o\", color=col, label=fr\"$\\it{{{g}}}$\")\n",
    "\n",
    "ax.set_ylabel(\"mean logCPM\")\n",
    "ax.set_title(f\"{model} — Astrocyte: Top variable genes\")\n",
    "ax.legend(bbox_to_anchor=(1.05,1), loc=\"upper left\")\n",
    "plt.xticks(rotation=45)\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "daf3c4a0-2ba5-4c38-9f9a-15bb6a267d41",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "\n",
    "model = \"MOG\"\n",
    "cm, var = variable_genes_per_model(\n",
    "    adata, model_name=model, MODELS=MODELS,\n",
    "    celltype=\"Astrocyte\"\n",
    ")\n",
    "\n",
    "top_genes = var_sub.head(5).index\n",
    "palette = sns.color_palette(\"tab10\", n_colors=len(top_genes))\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(8,5))\n",
    "for g, col in zip(top_genes, palette):\n",
    "    ax.plot(cm.columns, cm.loc[g], marker=\"o\", color=col, label=fr\"$\\it{{{g}}}$\")\n",
    "\n",
    "ax.set_ylabel(\"mean logCPM\")\n",
    "ax.set_title(f\"{model} — Astrocyte: Top variable genes\")\n",
    "ax.legend(bbox_to_anchor=(1.05,1), loc=\"upper left\")\n",
    "plt.xticks(rotation=45)\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e9f2663e-73c9-4453-99fb-3707c41bafa3",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "\n",
    "model = \"PLP\"\n",
    "cm, var = variable_genes_per_model(\n",
    "    adata, model_name=model, MODELS=MODELS,\n",
    "    celltype=\"Astrocyte\"\n",
    ")\n",
    "\n",
    "top_genes = var.head(5).index\n",
    "palette = sns.color_palette(\"tab10\", n_colors=len(top_genes))\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(8,5))\n",
    "for g, col in zip(top_genes, palette):\n",
    "    ax.plot(cm.columns, cm.loc[g], marker=\"o\", color=col, label=fr\"$\\it{{{g}}}$\")\n",
    "\n",
    "ax.set_ylabel(\"mean logCPM\")\n",
    "ax.set_title(f\"{model} — Astrocyte: Top variable genes\")\n",
    "ax.legend(bbox_to_anchor=(1.05,1), loc=\"upper left\")\n",
    "plt.xticks(rotation=45)\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7a770bc3-ba29-47fc-ba74-231c206e7176",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "\n",
    "model = \"PLP\"\n",
    "cm, var = variable_genes_per_model(\n",
    "    adata, model_name=model, MODELS=MODELS,\n",
    "    celltype=\"Astrocyte\"\n",
    ")\n",
    "\n",
    "top_genes = var.head(5).index\n",
    "palette = sns.color_palette(\"tab10\", n_colors=len(top_genes))\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(8,5))\n",
    "for g, col in zip(top_genes, palette):\n",
    "    ax.plot(cm.columns, cm.loc[g], marker=\"o\", color=col, label=fr\"$\\it{{{g}}}$\")\n",
    "\n",
    "ax.set_ylabel(\"mean logCPM\")\n",
    "ax.set_title(f\"{model} — Astrocyte: Top variable genes\")\n",
    "ax.legend(bbox_to_anchor=(1.05,1), loc=\"upper left\")\n",
    "plt.xticks(rotation=45)\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "78a59f05-fce2-45c5-b170-1218e1d0f9d4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>course</th>\n",
       "      <th>MOG CFA</th>\n",
       "      <th>PLP CFA</th>\n",
       "      <th>chronic long</th>\n",
       "      <th>chronic peak</th>\n",
       "      <th>early onset</th>\n",
       "      <th>monophasic</th>\n",
       "      <th>non symptomatic</th>\n",
       "      <th>onset I</th>\n",
       "      <th>onset II</th>\n",
       "      <th>peak I</th>\n",
       "      <th>peak II</th>\n",
       "      <th>peak III</th>\n",
       "      <th>remitt I</th>\n",
       "      <th>remitt II</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2610035D17Rik</th>\n",
       "      <td>0.760219</td>\n",
       "      <td>0.567049</td>\n",
       "      <td>0.684663</td>\n",
       "      <td>0.702827</td>\n",
       "      <td>0.600435</td>\n",
       "      <td>0.491483</td>\n",
       "      <td>0.641677</td>\n",
       "      <td>0.303612</td>\n",
       "      <td>0.386249</td>\n",
       "      <td>0.503658</td>\n",
       "      <td>0.420378</td>\n",
       "      <td>0.382122</td>\n",
       "      <td>0.484963</td>\n",
       "      <td>0.503577</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9630013A20Rik</th>\n",
       "      <td>0.040635</td>\n",
       "      <td>0.034014</td>\n",
       "      <td>0.064848</td>\n",
       "      <td>0.105693</td>\n",
       "      <td>0.032976</td>\n",
       "      <td>0.047670</td>\n",
       "      <td>0.027849</td>\n",
       "      <td>0.026098</td>\n",
       "      <td>0.049895</td>\n",
       "      <td>0.086409</td>\n",
       "      <td>0.072568</td>\n",
       "      <td>0.058821</td>\n",
       "      <td>0.078464</td>\n",
       "      <td>0.009566</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1cf</th>\n",
       "      <td>0.001182</td>\n",
       "      <td>0.004700</td>\n",
       "      <td>0.004409</td>\n",
       "      <td>0.007733</td>\n",
       "      <td>0.003580</td>\n",
       "      <td>0.001271</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.004661</td>\n",
       "      <td>0.003151</td>\n",
       "      <td>0.003579</td>\n",
       "      <td>0.002468</td>\n",
       "      <td>0.000658</td>\n",
       "      <td>0.004619</td>\n",
       "      <td>0.002409</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2m</th>\n",
       "      <td>5.090747</td>\n",
       "      <td>3.781237</td>\n",
       "      <td>5.922780</td>\n",
       "      <td>7.305157</td>\n",
       "      <td>5.906674</td>\n",
       "      <td>5.468080</td>\n",
       "      <td>4.578372</td>\n",
       "      <td>4.774786</td>\n",
       "      <td>4.965225</td>\n",
       "      <td>6.397564</td>\n",
       "      <td>5.945568</td>\n",
       "      <td>7.025800</td>\n",
       "      <td>5.428105</td>\n",
       "      <td>4.825909</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Aatf</th>\n",
       "      <td>0.765351</td>\n",
       "      <td>0.711719</td>\n",
       "      <td>0.837476</td>\n",
       "      <td>0.785777</td>\n",
       "      <td>0.729061</td>\n",
       "      <td>0.837987</td>\n",
       "      <td>0.645680</td>\n",
       "      <td>0.717868</td>\n",
       "      <td>0.790823</td>\n",
       "      <td>0.968513</td>\n",
       "      <td>0.869408</td>\n",
       "      <td>0.959560</td>\n",
       "      <td>0.804610</td>\n",
       "      <td>0.856146</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Zswim9</th>\n",
       "      <td>0.242324</td>\n",
       "      <td>0.259676</td>\n",
       "      <td>0.264105</td>\n",
       "      <td>0.247415</td>\n",
       "      <td>0.219694</td>\n",
       "      <td>0.250217</td>\n",
       "      <td>0.235258</td>\n",
       "      <td>0.186908</td>\n",
       "      <td>0.249990</td>\n",
       "      <td>0.231577</td>\n",
       "      <td>0.241121</td>\n",
       "      <td>0.266640</td>\n",
       "      <td>0.257441</td>\n",
       "      <td>0.274791</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Zup1</th>\n",
       "      <td>0.520665</td>\n",
       "      <td>0.338158</td>\n",
       "      <td>0.507997</td>\n",
       "      <td>0.442982</td>\n",
       "      <td>0.425951</td>\n",
       "      <td>0.342797</td>\n",
       "      <td>0.479227</td>\n",
       "      <td>0.235218</td>\n",
       "      <td>0.336152</td>\n",
       "      <td>0.374601</td>\n",
       "      <td>0.251204</td>\n",
       "      <td>0.284319</td>\n",
       "      <td>0.313890</td>\n",
       "      <td>0.355295</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Zyx</th>\n",
       "      <td>0.724900</td>\n",
       "      <td>0.723223</td>\n",
       "      <td>0.928527</td>\n",
       "      <td>1.042805</td>\n",
       "      <td>0.882195</td>\n",
       "      <td>0.884206</td>\n",
       "      <td>0.577124</td>\n",
       "      <td>0.703261</td>\n",
       "      <td>0.993592</td>\n",
       "      <td>1.086883</td>\n",
       "      <td>0.734526</td>\n",
       "      <td>1.178293</td>\n",
       "      <td>0.839119</td>\n",
       "      <td>0.772139</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Zzef1</th>\n",
       "      <td>1.127073</td>\n",
       "      <td>1.026866</td>\n",
       "      <td>1.124169</td>\n",
       "      <td>1.306193</td>\n",
       "      <td>1.169622</td>\n",
       "      <td>0.995753</td>\n",
       "      <td>1.033505</td>\n",
       "      <td>0.760119</td>\n",
       "      <td>1.002104</td>\n",
       "      <td>1.230116</td>\n",
       "      <td>0.939861</td>\n",
       "      <td>1.169605</td>\n",
       "      <td>1.056981</td>\n",
       "      <td>0.979196</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>a</th>\n",
       "      <td>0.001194</td>\n",
       "      <td>0.007333</td>\n",
       "      <td>0.005062</td>\n",
       "      <td>0.007838</td>\n",
       "      <td>0.008762</td>\n",
       "      <td>0.002403</td>\n",
       "      <td>0.003127</td>\n",
       "      <td>0.001436</td>\n",
       "      <td>0.004461</td>\n",
       "      <td>0.006138</td>\n",
       "      <td>0.001791</td>\n",
       "      <td>0.003443</td>\n",
       "      <td>0.007398</td>\n",
       "      <td>0.002275</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5101 rows × 14 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "course          MOG CFA   PLP CFA  chronic long  chronic peak  early onset  \\\n",
       "2610035D17Rik  0.760219  0.567049      0.684663      0.702827     0.600435   \n",
       "9630013A20Rik  0.040635  0.034014      0.064848      0.105693     0.032976   \n",
       "A1cf           0.001182  0.004700      0.004409      0.007733     0.003580   \n",
       "A2m            5.090747  3.781237      5.922780      7.305157     5.906674   \n",
       "Aatf           0.765351  0.711719      0.837476      0.785777     0.729061   \n",
       "...                 ...       ...           ...           ...          ...   \n",
       "Zswim9         0.242324  0.259676      0.264105      0.247415     0.219694   \n",
       "Zup1           0.520665  0.338158      0.507997      0.442982     0.425951   \n",
       "Zyx            0.724900  0.723223      0.928527      1.042805     0.882195   \n",
       "Zzef1          1.127073  1.026866      1.124169      1.306193     1.169622   \n",
       "a              0.001194  0.007333      0.005062      0.007838     0.008762   \n",
       "\n",
       "course         monophasic  non symptomatic   onset I  onset II    peak I  \\\n",
       "2610035D17Rik    0.491483         0.641677  0.303612  0.386249  0.503658   \n",
       "9630013A20Rik    0.047670         0.027849  0.026098  0.049895  0.086409   \n",
       "A1cf             0.001271         0.000000  0.004661  0.003151  0.003579   \n",
       "A2m              5.468080         4.578372  4.774786  4.965225  6.397564   \n",
       "Aatf             0.837987         0.645680  0.717868  0.790823  0.968513   \n",
       "...                   ...              ...       ...       ...       ...   \n",
       "Zswim9           0.250217         0.235258  0.186908  0.249990  0.231577   \n",
       "Zup1             0.342797         0.479227  0.235218  0.336152  0.374601   \n",
       "Zyx              0.884206         0.577124  0.703261  0.993592  1.086883   \n",
       "Zzef1            0.995753         1.033505  0.760119  1.002104  1.230116   \n",
       "a                0.002403         0.003127  0.001436  0.004461  0.006138   \n",
       "\n",
       "course          peak II  peak III  remitt I  remitt II  \n",
       "2610035D17Rik  0.420378  0.382122  0.484963   0.503577  \n",
       "9630013A20Rik  0.072568  0.058821  0.078464   0.009566  \n",
       "A1cf           0.002468  0.000658  0.004619   0.002409  \n",
       "A2m            5.945568  7.025800  5.428105   4.825909  \n",
       "Aatf           0.869408  0.959560  0.804610   0.856146  \n",
       "...                 ...       ...       ...        ...  \n",
       "Zswim9         0.241121  0.266640  0.257441   0.274791  \n",
       "Zup1           0.251204  0.284319  0.313890   0.355295  \n",
       "Zyx            0.734526  1.178293  0.839119   0.772139  \n",
       "Zzef1          0.939861  1.169605  1.056981   0.979196  \n",
       "a              0.001791  0.003443  0.007398   0.002275  \n",
       "\n",
       "[5101 rows x 14 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "course_means.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "71d9ed33-081f-4e50-9cb7-21bf1da1fad7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>var</th>\n",
       "      <th>range</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Cd74</th>\n",
       "      <td>4.811892e+00</td>\n",
       "      <td>6.307446</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>H2-D1</th>\n",
       "      <td>4.285040e+00</td>\n",
       "      <td>5.773047</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Serping1</th>\n",
       "      <td>3.851166e+00</td>\n",
       "      <td>6.077000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Serpina3n</th>\n",
       "      <td>3.560114e+00</td>\n",
       "      <td>6.332653</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>H2-K1</th>\n",
       "      <td>3.211678e+00</td>\n",
       "      <td>5.520392</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Myh7</th>\n",
       "      <td>8.202805e-07</td>\n",
       "      <td>0.003130</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tbpl2</th>\n",
       "      <td>7.894877e-07</td>\n",
       "      <td>0.002742</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gsx2</th>\n",
       "      <td>6.465283e-07</td>\n",
       "      <td>0.002480</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Hand1</th>\n",
       "      <td>5.580636e-07</td>\n",
       "      <td>0.001907</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Prdm14</th>\n",
       "      <td>5.165222e-07</td>\n",
       "      <td>0.002506</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5101 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    var     range\n",
       "Cd74       4.811892e+00  6.307446\n",
       "H2-D1      4.285040e+00  5.773047\n",
       "Serping1   3.851166e+00  6.077000\n",
       "Serpina3n  3.560114e+00  6.332653\n",
       "H2-K1      3.211678e+00  5.520392\n",
       "...                 ...       ...\n",
       "Myh7       8.202805e-07  0.003130\n",
       "Tbpl2      7.894877e-07  0.002742\n",
       "Gsx2       6.465283e-07  0.002480\n",
       "Hand1      5.580636e-07  0.001907\n",
       "Prdm14     5.165222e-07  0.002506\n",
       "\n",
       "[5101 rows x 2 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "variability"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "92ea9571-b189-48aa-be6c-d3e91bccc74c",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "sc",
   "language": "python",
   "name": "sc"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.20"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
